numpyro/infer/hmc_gibbs.py

```diff
@@ -0,0 +1,143 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import namedtuple
+import copy
+from functools import partial
+
+from jax import device_put, random, value_and_grad
+
+from numpyro.handlers import condition, seed, trace, substitute
+from numpyro.infer.mcmc import MCMCKernel
+from numpyro.infer.hmc import HMC
+
+
+HMCGibbsState = namedtuple("HMCGibbsState", "z, hmc_state, rng_key")
+"""
+ - **z** - a dict of the current latent values (both HMC and Gibbs sites)
+ - **hmc_state** - current hmc_state
+ - **rng_key** - random key for the current step
+"""
+
+
+def _wrap_model(model):
+    def fn(*args, **kwargs):
+        gibbs_values = kwargs.pop("_gibbs_sites", {})
+        with condition(data=gibbs_values), substitute(data=gibbs_values):
+            model(*args, **kwargs)
+    return fn
+
+
+class HMCGibbs(MCMCKernel):
+    """
+    [EXPERIMENTAL INTERFACE]
+
+    HMC-within-Gibbs. This inference algorithm allows the user to combine
+    general purpose gradient-based inference (HMC or NUTS) with custom
+    Gibbs samplers.
+
+    Note that it is the user's responsibility to provide a correct implementation
+    of `gibbs_fn` that samples from the corresponding posterior conditional.
+
+    :param inner_kernel: One of :class:`~numpyro.infer.HMC` or :class:`~numpyro.infer.NUTS`.
+    :param gibbs_fn: A Python callable that returns a dictionary of Gibbs samples conditioned
+        on the HMC sites. Must include an argument `rng_key` that should be used for all sampling.
+        Must also include arguments `hmc_sites` and `gibbs_sites`, each of which is a dictionary
+        with keys that are site names and values that are sample values. Note that a given `gibbs_fn`
+        may not need make use of all these sample values.
+    :param gibbs_sites: a list of site names for the latent variables that are covered by the Gibbs sampler.
+
+    **Example**
+
+    .. doctest::
+
+    >>> from jax import random
```

Comment:
hope this is correct formatting!

Reply:
I think you will need to indent here (not so sure about that). You can test the docstring with `make doctest`.